test/MathUtils.test.cpp

```diff
@@ -0,0 +1,237 @@
+	EXPECT_EQ(1, NAS2D::divideUp(5, 8));
+	EXPECT_DEATH(NAS2D::divideUp(256, 0), "divideUp(256,0) divide by zero.");
+	EXPECT_EQ(256, NAS2D::divideUp(256, 1));
+	EXPECT_EQ(128, NAS2D::divideUp(256, 2));
+	EXPECT_EQ(86, NAS2D::divideUp(256, 3));
+	EXPECT_EQ(64, NAS2D::divideUp(256, 4));
+	EXPECT_EQ(52, NAS2D::divideUp(256, 5));
+	EXPECT_EQ(44, NAS2D::divideUp(256, 6));
+	EXPECT_EQ(38, NAS2D::divideUp(256, 7));
+	EXPECT_EQ(32, NAS2D::divideUp(256, 8));
+	EXPECT_EQ(29, NAS2D::divideUp(256, 9));
+	EXPECT_EQ(26, NAS2D::divideUp(256, 10));
+	EXPECT_EQ(24, NAS2D::divideUp(256, 11));
+	EXPECT_EQ(22, NAS2D::divideUp(256, 12));
+	EXPECT_EQ(20, NAS2D::divideUp(256, 13));
+}
+
+TEST(MathUtils, isPointInRectComponent)
+{
+	//Neghbors
+	EXPECT_EQ(false, NAS2D::isPointInRect(4, 4, 5, 5, 1, 1));
+	EXPECT_EQ(false, NAS2D::isPointInRect(5, 4, 5, 5, 1, 1));
+	EXPECT_EQ(false, NAS2D::isPointInRect(6, 4, 5, 5, 1, 1));
+	EXPECT_EQ(false, NAS2D::isPointInRect(7, 4, 5, 5, 1, 1));
+	EXPECT_EQ(false, NAS2D::isPointInRect(7, 5, 5, 5, 1, 1));
+	EXPECT_EQ(false, NAS2D::isPointInRect(7, 6, 5, 5, 1, 1));
+	EXPECT_EQ(false, NAS2D::isPointInRect(7, 7, 5, 5, 1, 1));
+	EXPECT_EQ(false, NAS2D::isPointInRect(6, 7, 5, 5, 1, 1));
+	EXPECT_EQ(false, NAS2D::isPointInRect(5, 7, 5, 5, 1, 1));
+	EXPECT_EQ(false, NAS2D::isPointInRect(4, 7, 5, 5, 1, 1));
+	EXPECT_EQ(false, NAS2D::isPointInRect(4, 6, 5, 5, 1, 1));
+	EXPECT_EQ(false, NAS2D::isPointInRect(4, 5, 5, 5, 1, 1));
+	EXPECT_EQ(false, NAS2D::isPointInRect(4, 4, 5, 5, 1, 1));
+
+	//Top
+	EXPECT_EQ(false, NAS2D::isPointInRect(4, 5, 5, 5, 1, 1));
+	EXPECT_EQ(true, NAS2D::isPointInRect(5, 5, 5, 5, 1, 1));
+	EXPECT_EQ(true, NAS2D::isPointInRect(6, 5, 5, 5, 1, 1));
+	EXPECT_EQ(false, NAS2D::isPointInRect(7, 5, 5, 5, 1, 1));
+
+	//Bottom
+	EXPECT_EQ(false, NAS2D::isPointInRect(4, 6, 5, 5, 1, 1));
+	EXPECT_EQ(true, NAS2D::isPointInRect(5, 6, 5, 5, 1, 1));
+	EXPECT_EQ(true, NAS2D::isPointInRect(6, 6, 5, 5, 1, 1));
+	EXPECT_EQ(false, NAS2D::isPointInRect(7, 6, 5, 5, 1, 1));
+
+	//Left
+	EXPECT_EQ(false, NAS2D::isPointInRect(5, 4, 5, 5, 1, 1));
+	EXPECT_EQ(true, NAS2D::isPointInRect(5, 5, 5, 5, 1, 1));
+	EXPECT_EQ(true, NAS2D::isPointInRect(5, 6, 5, 5, 1, 1));
+	EXPECT_EQ(false, NAS2D::isPointInRect(5, 7, 5, 5, 1, 1));
+
+	//Right
+	EXPECT_EQ(false, NAS2D::isPointInRect(6, 4, 5, 5, 1, 1));
+	EXPECT_EQ(true, NAS2D::isPointInRect(6, 5, 5, 5, 1, 1));
+	EXPECT_EQ(true, NAS2D::isPointInRect(6, 6, 5, 5, 1, 1));
+	EXPECT_EQ(false, NAS2D::isPointInRect(6, 7, 5, 5, 1, 1));
+
+	//Center
+	EXPECT_EQ(false, NAS2D::isPointInRect(4, 6, 5, 5, 2, 2));
+	EXPECT_EQ(true, NAS2D::isPointInRect(5, 6, 5, 5, 2, 2));
+	EXPECT_EQ(true, NAS2D::isPointInRect(6, 6, 5, 5, 2, 2));
+	EXPECT_EQ(true, NAS2D::isPointInRect(7, 6, 5, 5, 2, 2));
+	EXPECT_EQ(false, NAS2D::isPointInRect(8, 6, 5, 5, 2, 2));
+}
+
+TEST(MathUtils, isPointInRectClass)
+{
+	using namespace NAS2D;
+	//Neghbors
+	EXPECT_EQ(false, NAS2D::isPointInRect({4, 4}, {5, 5, 1, 1}));
```

Comment:
This set of tests is much easier to read than the one above. Even though individual values are passed here, the `{}` adds minimal syntactic noise, and yet makes it much easier to read and understand the code at the point of call.

As the test here shows what is possible, I think I'd recommend dropping the other variant of the method that allows passing unpacked values.

From an assembly level perspective, I don't believe the unpacked values would have any efficiency advantage over the packed values. Particularly not if the structs are passed by value, as it would all just be pushed onto the stack either way.

Reply:
I think using explicit class names makes it even more clear. Should be the programmer's choice though. Maybe we should provide examples of both syntax in the documentation.

Reply:
I think I'm starting to agree more with explicit constructors. At the very least, I think example code should use the type names when using initializer lists.